### PR TITLE
Hexagon dma Transfer Rework patch-1

### PR DESCRIPTION
**Pull request by vgundlur**
Halide UBWC/DMA test function, for step 1 of 5
    Step 1: Simple, synchronous, DMA tiling example, using cross buffer copy. Let's *not* attempt to share resources across pipeline invocations, even if it's slow for now. It will be easier to figure out how to share resources across pipeline invocations once we can see exactly what needs to be shared and how.
1) Codeset partitioned into frontend and backend, with the following layering
   Halide-app -> nhalide_pipeline -> frontend ->adaptor -> backend
   Where
    a) mhalide_pipeline: pseudo testcode
    b) frontend: the new APIs as specified for each step’s pseudocode
    c) adaptor: the layer of code that adapts to make the frontend work with backend
    d) backend: (leveraging where possible) existing code to speedup code turnaround time
   Frontend represented by:
    a) halide_hexagon_dma_api.h/cpp
   Adaptor represented by:
    b) hexagon_dma_api.h/cpp
    
The plan is for frontend code to lead backend, with backend to progressively align thru the steps
1. Key changes
    a) Decoupled/separation of dma_context from user_context
    b) Introduced context class, using OpenCL/CUDA approach
    c) Use of halide_buffer_t’s device field for dma_context
    d) Swap host (locked L2$) and device (DDR) roles
    e) Single frame of NV12, processing luma only
    f) Discrepancies from original pseudocode
        i)  Using default parameters and hardcoding (see comments prefixed with // ASSUMPTION:        
2) check algo for deriving 420 pixel component/plane, from halide_dimention_t
    a) See halide_hexagon_dma_comp_get()

### CODE REVIEW

File: src/runtime/hexagon_dma_device_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+namespace Halide { namespace Runtime { namespace Internal { namespace Dma {
+
+//Helper class to pass DMA Context
+class HexagonDmaContext {
+    void* user_ctxt;
+
+public:
+    dma_context context;
```
Comment (pranavb-ca) [inline src/runtime/hexagon_dma_device_interface.cpp]: Any reason you have a public data member?

File: src/runtime/hexagon_dma_api.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+typedef struct {
+   uintptr_t fold_vaddr;
+   int fold_idx;
+} cache_mem_t;
+
+/**
+ * DmaContext
+ * global type of dma context
+ */
+typedef DmaContext* dma_context;
```
Comment (pranavb-ca) [inline src/runtime/hexagon_dma_api.h]: Minor nitpick: please consider changing this to
`typedef DmaContext * dma_context_t;`

For some reason, dma_context feels like you are referring to an object, but whereas in fact we are dealign with a pointer.

File: src/runtime/HalideRuntimeHexagonDma.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * halide_buffer_copy
+ * in:void* user_Context
+ * out:halide_buffer_t src
+ * out:halide_buffer_t dest
+ * void* ptr */
+int halide_buffer_copy(void *user_context, halide_buffer_t *src_buf, void *ptr, halide_buffer_t *dst_buf);
+
+/**
+ * halide_hexagon_dmaapp_create_context
+ * CreateContext check if DMA Engines are available
```
Comment (ronlieb) [inline src/runtime/HalideRuntimeHexagonDma.h]: Most of these comment blocks for the prototypes seem redundant, they just take up space and repeat what is already apparent in the prototype. I would suggest replacing the content with a meaningful statement about the function, or removing it all together.

File: src/runtime/hexagon_dma_api.cpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        nRet  = dma_lock_cache(tcm_desc_vaddr, region_tcm_desc_size);
+        if (nRet != QURT_EOK) {
+            error(user_context) << "QURT TCM lock failed due to QURT_EALIGN ERROR  \
+                       misaligned u32Size = " << region_tcm_desc_size << "\n" ;
+            return NULL;
+        }
+    }
+
+    t_mem_desc* cache_mem = (t_mem_desc *) malloc (sizeof(t_mem_desc));
+ 	cache_mem->desc_region = region_tcm_desc;
```
Comment (dpalermo) [inline src/runtime/hexagon_dma_api.cpp]: line up indentation with code below

File: src/runtime/hexagon_dma_api.cpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        }
+        return NULL;
+    }
+
+    free(cache_mem);
+    return (void*)tcm_buf_vaddr;
+}
+
+int  halide_hexagon_dma_wrap_memory_free(void* user_context, dma_context pdma_context, void* vret) {
+	  if (vret == NULL) {
```
Comment (dpalermo) [inline src/runtime/hexagon_dma_api.cpp]: line up indentation with code below...entire function actually needs indent shifted left

File: src/runtime/hexagon_dma_api.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+typedef enum {
+    HEX_SUCCESS = 0,
+    HEX_ERROR = -1
+} t_dma_error_codes;
+
+/**
+ * DmaContext
+ * global type of dma context
+ */
+typedef struct t_dma_context* dma_context;
```
Comment (pranavb-ca) [inline src/runtime/hexagon_dma_api.h]: Please consider changing the name here.
dma_context makes me think that you are referring to an object of class t_dma_context, but not to a pointer. Please consider dma_context_t or some other such name.

File: src/runtime/hexagon_dma_api.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace Halide { namespace Runtime { namespace Internal { namespace Dma {
+
+//Helper class to pass DMA Context
+class HexagonDmaContext {
+    void* user_ctxt;
+
+public:
+    dma_context context;
```
Comment (pranavb-ca) [inline src/runtime/hexagon_dma_api.h]: Does this need to be a public member?

File: src/runtime/halide_hexagon_dma_api.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    halide_assert(user_context, buf->dimensions < 4);
+
+    int num_of_frames = 1;  // ASSUMPTION: step 1 is one input frame only
+    int last_frame = 1;     // ASSUMPTION: always new resources for each frame
+
+    HexagonDmaContext hexagon_dma_context(user_context, num_of_frames);
+    halide_assert(user_context, hexagon_dma_context.context != NULL);
+    dma_context dma_ctxt = hexagon_dma_context.context;
+
+    halide_hexagon_dmart_attach_context(user_context, dma_ctxt, (uintptr_t)inframe, fmt, read, \
```
Comment (pranavb-ca) [inline src/runtime/halide_hexagon_dma_api.cpp]: This function returns an error code. Please catch the error code and return it from here.

File: src/runtime/halide_hexagon_dma_api.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+/**
+ * halide_hexagon_dmart_get_memory
+ * desc: Calculate width, height, stride and fold from roi buf
+ * Need also type and padding information. How ???
+ * Allocate Memory
+ */
+void* halide_hexagon_dmaapp_get_memory(void* user_context, halide_buffer_t *roi_buf) {
+    void *vret = halide_hexagon_dma_memory_get(user_context, roi_buf);
```
Comment (pranavb-ca) [inline src/runtime/halide_hexagon_dma_api.cpp]: Do you think this function will be called from anywhere else? At this point i see that this is the only call site for halide_hexagon_dma_memory_get. If this is going to be the only call-site please consider inlining halide_hexagon_dma_memory_get here. It'll be easier to see what's going on.

File: src/runtime/halide_hexagon_dma_api.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    buf->device_interface = NULL;
+
+    return HEX_SUCCESS;
+}
+
+/**
+ * halide_hexagon_dmart_get_memory
+ * desc: Calculate width, height, stride and fold from roi buf
+ * Need also type and padding information. How ???
+ * Allocate Memory
```
Comment (pranavb-ca) [inline src/runtime/halide_hexagon_dma_api.cpp]: Please mention what time of memory is being allocated in the comment here.

File: src/runtime/halide_hexagon_dma_api.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        /////////////////////////////////////////////////////////
+        //ASSUMPTION: Assuming paddng and fmt type to be default here since we really do  not have the inframe
+        //int halide_hexagon_dmart_get_padding(void* user_context, uintptr_t frame);
+        //int halide_hexagon_dmart_get_fmt_type(void* user_context, uintptr_t frame)
+        ///////////////////////////////////////////////////
+        int padding = 0;
+        int type = NV12;
+
+        // allocate L2$
+        vret = halide_hexagon_dma_memory_alloc(user_context, dma_ctxt, comp, w, h, s, nCircularFold, padding, type);
```
Comment (pranavb-ca) [inline src/runtime/halide_hexagon_dma_api.cpp]: Please consider rewriting this as
`vret = halide_hexagon_dma_memory_alloc(user_context, dma_ctxt, comp, roi_buf->dim[0].extent,
                                                                          roi_buf->dim[1].extent, roi_buf->dim[1].stride,
                                                                           nCircularFold, padding, type);`

This makes it easy to see what is being passed.

File: src/runtime/halide_hexagon_dma_api.cpp (hunk 9 above)
Comment (pranavb-ca) [reply src/runtime/halide_hexagon_dma_api.cpp]: Also consider returning early if vret is not null. So consider changing this code to
` halide_assert(user_context, roi_buf != NULL);
   if(roi_buf->host != 0) {
        return (void*) roi_buf->host;
    } else {
     }`

**Review by pranavb-ca — CHANGES_REQUESTED** [review]

**Comment by vgundlur** [comment]
Internal review done, will open a PR on Master Branch

**Pull request closed without merging**